server/controllers/nlp.js

```diff
@@ -17,19 +29,210 @@ let functions = {
+            state.tokens[state.currentToken] = {type: static.column, value: currentToken};
+            matched = true;
+        } else {
+            //look for synonyms
+            _.forEach(columnSynonyms, function (column) {
+                if(_.includes(column.synomyms, currentToken)){
+                    tokenMatched  = column.columnName;
+                    return false;
+                }
+            });
+            if(tokenMatched ){
+                state.tokens[state.currentToken] = {type: static.column, value: tokenMatched }
+                matched = true;
+            }else{
+                _.forEach(columnSynonyms, function (column) {
+                    var firstTokenMatched  = _.filter(column.synomyms, function (o) {
+                        var firstWord = o.replace(/ .*/, '');
+                        return firstWord === currentToken;
+                    });
+                    if(firstTokenMatched .length > 0){
+                        //look if the second matches
+                        _.forEach(firstTokenMatched , function (item) {
+                            var words = item.split(' ');
+                            if (words[1] === state.tokens[state.currentToken + 1]) {
+                                //replace the currentToken
+                                state.tokens[state.currentToken] = {type: static.column, value: column.columnName}
+                                state.tokens.splice(state.currentToken + 1, 1);
+                                matched = true;
+                            }
+                        });
+                    }
+                });
+            }
+        }
+        nextAction(matched, state);
+    });
+}
+
+function extractChartType(state) {
+    var currentToken = state.tokens[state.currentToken]
+    let possibleTypes = ["histogram", "pie chart", "line chart", "bar chart", "scatter plot"];
+    let matched = false;
+    var tokenMatched  = _.includes(possibleTypes, currentToken);
+
+    if (tokenMatched ) {
+        // this layer matches for this token on single word
+        state.tokens[state.currentToken] = {type: static.chartType, value: currentToken}
+        matched = true;
+    } else {
+        //search if the first word of a Type matches
+        tokenMatched  = _.find(possibleTypes, function (o) {
+            var firstWord = o.replace(/ .*/, '');
+            return firstWord === currentToken;
+        });
+        if (tokenMatched ) {
+            //look if the second matches
+            var words = tokenMatched .split(' ');
+            if (words[1] === state.tokens[state.currentToken + 1]) {
+                //replace the currentToken
+                state.tokens[state.currentToken] = {type: static.chartType, value: tokenMatched }
+                state.tokens.splice(state.currentToken + 1, 1);
+                matched = true;
+            }
+        }
+    }
+    nextAction(matched, state);
+}
+
+function extractConnector(state) {
```

Comment:
this can be removed for now, right?

Reply:
i thought we want to extract the connectors also later, so i just left it in, but ok